File: monitors/eyes.py
"""
ตรวจดวงตา: นับการกะพริบ + ตรวจความง่วง (ตาปิดค้าง)

ใช้ค่า EAR (Eye Aspect Ratio) — อัตราส่วนความสูงต่อความกว้างของดวงตา
ตาเปิด EAR สูง / ตาหรี่หรือปิด EAR ต่ำ

ประโยชน์:
  - อัตรากะพริบต่ำผิดปกติ (< ~8 ครั้ง/นาที) = เพ่งจอนาน เสี่ยงตาแห้ง
  - ตาปิดค้างนาน = ง่วง/หลับใน (PERCLOS)
"""
from collections import deque

import numpy as np

# ดัชนี landmark ของ MediaPipe FaceMesh สำหรับจุดรอบดวงตา (6 จุดต่อข้าง)
LEFT_EYE = [33, 160, 158, 133, 153, 144]
RIGHT_EYE = [362, 385, 387, 263, 373, 380]


def _eye_aspect_ratio(landmarks, idx, w, h) -> float:
    pts = [(landmarks[i].x * w, landmarks[i].y * h) for i in idx]
    p1, p2, p3, p4, p5, p6 = pts

    def dist(a, b):
        return float(np.hypot(a[0] - b[0], a[1] - b[1]))

    vertical = dist(p2, p6) + dist(p3, p5)   # ความสูงเปลือกตา (สองคู่)
    horizontal = 2.0 * dist(p1, p4)           # ความกว้างหัวตา-หางตา
    return vertical / horizontal if horizontal else 0.0


class EyeMonitor:
    def __init__(self, ear_threshold: float = 0.21, drowsy_seconds: float = 1.2):
        self.ear_threshold = ear_threshold     # ต่ำกว่านี้ถือว่าตาปิด
        self.drowsy_seconds = drowsy_seconds   # ปิดค้างนานเกินนี้ = ง่วง
        self.ear: float = 0.0
        self.blink_count: int = 0
        self.drowsy: bool = False

        self._eye_closed = False
        self._closed_start: float | None = None
        self._blink_times: deque[float] = deque()  # เก็บเวลากะพริบ 60 วินาทีล่าสุด

    def update(self, landmarks, w, h, timestamp: float) -> None:
        left = _eye_aspect_ratio(landmarks, LEFT_EYE, w, h)
        right = _eye_aspect_ratio(landmarks, RIGHT_EYE, w, h)
        self.ear = (left + right) / 2.0

        if self.ear < self.ear_threshold:
            # ตาปิดอยู่
            if not self._eye_closed:
                self._eye_closed = True
                self._closed_start = timestamp
            elif timestamp - self._closed_start >= self.drowsy_seconds:
                self.drowsy = True  # ปิดค้างนานเกินไป
        else:
            # ตาเปิด
            if self._eye_closed and self._closed_start is not None:
                closed_dur = timestamp - self._closed_start
                # ปิดสั้น ๆ แล้วเปิด = นับเป็นการกะพริบ 1 ครั้ง
                if closed_dur < self.drowsy_seconds:
                    self.blink_count += 1
                    self._blink_times.append(timestamp)
            self._eye_closed = False
            self.drowsy = False

        # คงไว้เฉพาะการกะพริบในช่วง 60 วินาทีล่าสุด
        while self._blink_times and timestamp - self._blink_times[0] > 60:
            self._blink_times.popleft()

    @property
    def blink_rate(self) -> int:
        """จำนวนครั้งที่กะพริบใน 60 วินาทีที่ผ่านมา"""
        return len(self._blink_times)

File: web/src/monitors/eyes.ts
/**
 * ตรวจดวงตา: นับการกะพริบ + ตรวจความง่วง (ตาปิดค้าง) — พอร์ตจาก monitors/eyes.py
 *
 * ใช้ค่า EAR (Eye Aspect Ratio) — อัตราส่วนความสูงต่อความกว้างของดวงตา
 * ตาเปิด EAR สูง / ตาหรี่หรือปิด EAR ต่ำ
 *
 * ประโยชน์:
 *   - อัตรากะพริบต่ำผิดปกติ (< ~8 ครั้ง/นาที) = เพ่งจอนาน เสี่ยงตาแห้ง
 *   - ตาปิดค้างนาน = ง่วง/หลับใน (PERCLOS)
 */
import type { NormalizedLandmark } from "@mediapipe/tasks-vision";

// ดัชนี landmark ของ MediaPipe FaceMesh สำหรับจุดรอบดวงตา (6 จุดต่อข้าง)
const LEFT_EYE = [33, 160, 158, 133, 153, 144];
const RIGHT_EYE = [362, 385, 387, 263, 373, 380];

/** ระยะทางแบบยุคลิด (Euclidean distance) ระหว่างจุดสองจุด */
function dist(a: [number, number], b: [number, number]): number {
  return Math.hypot(a[0] - b[0], a[1] - b[1]);
}

/** คำนวณค่า EAR จากจุด landmark 6 จุดของดวงตาหนึ่งข้าง */
function eyeAspectRatio(
  landmarks: NormalizedLandmark[],
  idx: number[],
  w: number,
  h: number,
): number {
  // แปลงพิกัด normalized (0..1) เป็นพิกเซลตามขนาดเฟรม
  const pts: [number, number][] = idx.map((i) => [landmarks[i].x * w, landmarks[i].y * h]);
  const [p1, p2, p3, p4, p5, p6] = pts;

  const vertical = dist(p2, p6) + dist(p3, p5); // ความสูงเปลือกตา (สองคู่)
  const horizontal = 2.0 * dist(p1, p4); // ความกว้างหัวตา-หางตา
  return horizontal ? vertical / horizontal : 0.0;
}

export class EyeMonitor {
  ear = 0.0;
  blinkCount = 0;
  drowsy = false;

  private eyeClosed = false;
  private closedStart: number | null = null;
  private blinkTimes: number[] = []; // เก็บเวลากะพริบ 60 วินาทีล่าสุด

  constructor(
    private earThreshold = 0.21, // ต่ำกว่านี้ถือว่าตาปิด
    private drowsySeconds = 1.2, // ปิดค้างนานเกินนี้ = ง่วง
  ) {}

  update(lm: NormalizedLandmark[], w: number, h: number, tSec: number): void {
    const left = eyeAspectRatio(lm, LEFT_EYE, w, h);
    const right = eyeAspectRatio(lm, RIGHT_EYE, w, h);
    this.ear = (left + right) / 2.0;

    if (this.ear < this.earThreshold) {
      // ตาปิดอยู่
      if (!this.eyeClosed) {
        this.eyeClosed = true;
        this.closedStart = tSec;
      } else if (this.closedStart !== null && tSec - this.closedStart >= this.drowsySeconds) {
        this.drowsy = true; // ปิดค้างนานเกินไป
      }
    } else {
      // ตาเปิด
      if (this.eyeClosed && this.closedStart !== null) {
        const closedDur = tSec - this.closedStart;
        // ปิดสั้น ๆ แล้วเปิด = นับเป็นการกะพริบ 1 ครั้ง
        if (closedDur < this.drowsySeconds) {
          this.blinkCount += 1;
          this.blinkTimes.push(tSec);
        }
      }
      this.eyeClosed = false;
      this.drowsy = false;
    }

    // คงไว้เฉพาะการกะพริบในช่วง 60 วินาทีล่าสุด
    while (this.blinkTimes.length && tSec - this.blinkTimes[0] > 60) {
      this.blinkTimes.shift();
    }
  }

  /** จำนวนกะพริบใน 60 วินาทีล่าสุด */
  get blinkRate(): number {
    return this.blinkTimes.length;
  }
}
